### PR TITLE
add public-viewer support to MemberClusterService

### PR DESCRIPTION
**Pull request by filariow**
This PR is part of the effort of splitting PR #443

Signed-off-by: Francesco Ilario <filario@redhat.com>

### CODE REVIEW

**Review by alexeykazakov — COMMENTED** [review]
Overall looks good (I haven't look at the tests yet)!

File: pkg/proxy/service/cluster_service_test.go [hunk 1]
```diff
@@ -107,320 +107,372 @@ func (s *TestClusterServiceSuite) TestGetClusterAccess() {
...
+							assert.Equal(s.T(), "smith2", ca.Username())
+						})
+					})
 				})
 			})
 		})
+	}
+
+	// public-viewer enabled
+	s.Run("user is public-viewer", func() {
```
Comment (mfrancisc) [inline pkg/proxy/service/cluster_service_test.go:452]: should we have test cases also with these workspaces:
- 456-not-ready
- 012-ready-unknown-cluster 

WDYT?

**Review by mfrancisc — COMMENTED** [review]
Overall looks good! Great job 👍

**Comment by mfrancisc** [comment]
One comment I have is that the unit tests are becoming more complicated to follow now, I wonder if we really need to check all the cases with the additional layers , or if we could just add new test cases targeting the public viewer scenarios and configure the existing tests with publicViewer set to `false`, but maybe that will leave with some untested scenarios.

File: pkg/proxy/service/cluster_service.go [hunk 2]
```diff
@@ -40,21 +41,21 @@ func NewMemberClusterService(context servicecontext.ServiceContext, options ...O
...
+		return s.getClusterAccessForDefaultWorkspace(userID, username, proxyPluginName)
+	}
+
+	return s.getSpaceAccess(userID, username, workspace, proxyPluginName, publicViewerEnabled)
+}
+
+// getSpaceAccess retrieves space access for an user
+func (s *ServiceImpl) getSpaceAccess(userID, username, workspace, proxyPluginName string, publicViewerEnabled bool) (*access.ClusterAccess, error) {
+	// retrieve the user's complaint name
+	userName, err := s.getUserSignupComplaintName(userID, username, publicViewerEnabled)
```
[the remark this answers is not in the document]
Comment (filariow) [reply pkg/proxy/service/cluster_service.go]: done in e66c362a247d4b34456fa41ccb80bfb30a262098

File: pkg/proxy/service/cluster_service.go [hunk 3]
```diff
@@ -65,7 +66,52 @@ func (s *ServiceImpl) GetClusterAccess(userID, username, workspace, proxyPluginN
 		return nil, fmt.Errorf("the requested space is not available")
 	}
 
-	return s.accessForSpace(space, signup.CompliantUsername, proxyPluginName)
+	return s.accessForSpace(space, userName, proxyPluginName)
```
[the remark this answers is not in the document]
Comment (filariow) [reply pkg/proxy/service/cluster_service.go]: done in e66c362a247d4b34456fa41ccb80bfb30a262098

File: pkg/proxy/service/cluster_service.go [hunk 4]
```diff
@@ -65,7 +66,52 @@ func (s *ServiceImpl) GetClusterAccess(userID, username, workspace, proxyPluginN
...
+}
+
+func (s *ServiceImpl) getUserSignupComplaintName(userID, username string, publicViewerEnabled bool) (string, error) {
+	// if PublicViewer is enabled and the requested user is the PublicViewer, than no lookup is required
+	if publicViewerEnabled && username == userID && userID == toolchainv1alpha1.KubesawAuthenticatedUsername {
```
[the remark this answers is not in the document]
Comment (filariow) [reply pkg/proxy/service/cluster_service.go]: done in b711d62e93ac9466e00dac83e721e9324bf8f0ae

File: pkg/proxy/service/cluster_service_test.go [hunk 5]
```diff
@@ -107,320 +107,372 @@ func (s *TestClusterServiceSuite) TestGetClusterAccess() {
...
-				return nil, errors.New("oopsi woopsi")
-			}
-
-			// when
-			_, err := svc.GetClusterAccess("789-ready", "", "", "")
-
-			// then
-			require.EqualError(s.T(), err, "oopsi woopsi")
-		})
+	tt := map[string]struct {
```
[the remark this answers is not in the document]
Comment (filariow) [reply pkg/proxy/service/cluster_service_test.go:110]: yes, leaving aside "public-viewer enabled/disabled" the following are the changes to the tests:

```diff
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_no_route/cluster_access_correct_when_username_provided
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_no_route/cluster_access_correct_when_using_workspace_context
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_no_route/cluster_access_correct_when_using_workspace_context/another_workspace_on_another_cluster
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_with_route/cluster_access_correct_when_username_provided
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_with_route/cluster_access_correct_when_using_workspace_context
TestRunClusterServiceSuite/TestGetClusterAccess/member_found/verify_cluster_access_with_route/cluster_access_correct_when_using_workspace_context/another_workspace_on_another_cluster
TestRunClusterServiceSuite/TestGetClusterAccess/no_member_cluster_found/no_member_cluster_with_the_given_URL/default_workspace_case
TestRunClusterServiceSuite/TestGetClusterAccess/no_member_cluster_found/no_member_cluster_with_the_given_URL/workspace_context_case
TestRunClusterServiceSuite/TestGetClusterAccess/no_member_cluster_found/no_member_clusters/default_workspace_case
TestRunClusterServiceSuite/TestGetClusterAccess/no_member_cluster_found/no_member_clusters/workspace_context_case
- TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/signup_service_returns_error
- TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/user_is_not_provisioned_yet
- TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/userid_is_not_found
- TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/username_is_not_found
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/default_workspace/signup_service_returns_error
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/default_workspace/user_is_not_provisioned_yet
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/default_workspace/userid_is_not_found
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/default_workspace/username_is_not_found
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/not-existing_workspace/signup_service_returns_error
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/not-existing_workspace/user_is_not_provisioned_yet
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/not-existing_workspace/userid_is_not_found
+TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_signup/not-existing_workspace/username_is_not_found
TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_space/informer_service_returns_error
TestRunClusterServiceSuite/TestGetClusterAccess/unable_to_get_space/space_not_found
```

File: pkg/proxy/service/cluster_service_test.go [hunk 6]
```diff
@@ -107,320 +107,372 @@ func (s *TestClusterServiceSuite) TestGetClusterAccess() {
...
+					Svcs:   s.Application,
+				},
+				func(si *service.ServiceImpl) {
+					si.GetMembersFunc = func(_ ...commoncluster.Condition) []*commoncluster.CachedToolchainCluster {
+						return s.memberClusters()
+					}
+				},
+			)
+			_, err := svc.GetClusterAccess(toolchainv1alpha1.KubesawAuthenticatedUsername, toolchainv1alpha1.KubesawAuthenticatedUsername, "smith2", "", true)
+			require.NoError(s.T(), err)
```
[the remark this answers is not in the document]
Comment (filariow) [reply pkg/proxy/service/cluster_service_test.go]: good point, done in a260aa3f5671226424a31e769e9abde598886829

**Comment by filariow** [comment]
> One comment I have is that the unit tests are becoming more complicated to follow now, I wonder if we really need to check all the cases with the additional layers , or if we could just add new test cases targeting the public viewer scenarios and configure the existing tests with publicViewer set to `false`, but maybe that will leave with some untested scenarios.

When public-viewer is enabled we expect the same behavior we have when public-viewer is disabled with a few exception for the KubesawAuthenticated user. So IMO to test it correctly we need to test the the same tests are passing with either public-viewer enabled and disabled. Unfortunately this complicates tests.

**Review by alexeykazakov — COMMENTED** [review]
Great job! And good test coverage. However if I not mistaken there is one test case missing: the public viewer access is disabled but the public viewer authenticated username is used.

**Comment by filariow** [comment]
> Great job! And good test coverage. However, if I not mistaken, there is one test case missing: the public viewer access is disabled but the public viewer authenticated username is used.

I looked into it and eventually decided to not add this test case because
1. `kube` is in the list of forbidden prefixes so it should never happen as `kubesaw-authenticated` matches it
2. even if it existed, it would be treated as a normal user (scenario already tested)

Btw, it should not be neither difficult nor harmful to have this test case. WDYT, should we add it anyway?

File: pkg/proxy/service/cluster_service.go [hunk 7]
```diff
@@ -65,7 +66,52 @@ func (s *ServiceImpl) GetClusterAccess(userID, username, workspace, proxyPluginN
 		return nil, fmt.Errorf("the requested space is not available")
 	}
 
-	return s.accessForSpace(space, signup.CompliantUsername, proxyPluginName)
+	return s.accessForSpace(space, complaintUserName, proxyPluginName)
+}
+
+func (s *ServiceImpl) getUserSignupComplaintName(userID, username string, publicViewerEnabled bool) (string, error) {
+	// if PublicViewer is enabled and the requested user is the PublicViewer, than no lookup is required
+	if publicViewerEnabled && username == toolchainv1alpha1.KubesawAuthenticatedUsername {
```
Comment (alexeykazakov) [inline pkg/proxy/service/cluster_service.go:74]: Will your tests fail if you replace this by
```suggestion
	if username == toolchainv1alpha1.KubesawAuthenticatedUsername {
```
?
If yes then we are good. If not then I would add a test case for this.

**Review by alexeykazakov — COMMENTED** [review]
> kube is in the list of forbidden prefixes so it should never happen as kubesaw-authenticated matches it
even if it existed, it would be treated as a normal user (scenario already tested)
Btw, it should not be neither difficult nor harmful to have this test case. WDYT, should we add it anyway?

OK. It makes sense. The only reason for such test would be to make sure that using `kubesaw-authenticated` would still require the public viewer feature to be enabled. See my comment below.

File: pkg/proxy/service/cluster_service_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mfrancisc) [reply pkg/proxy/service/cluster_service_test.go:110]: Ok Thanks for confirming!

File: pkg/proxy/service/cluster_service.go (hunk 7 above)
Comment (filariow) [reply pkg/proxy/service/cluster_service.go:74]: great point! that should be it: d9ba99c0ac11bcebf5b9c246090ff1e94c11e8dd

File: pkg/proxy/service/cluster_service_test.go [hunk 8]
```diff
@@ -107,318 +107,410 @@ func (s *TestClusterServiceSuite) TestGetClusterAccess() {
...
-			s.Run("cluster access correct when username provided", func() {
+			s.Run("public-viewer is disabled", func() {
 				// when
-				ca, err := svc.GetClusterAccess("", "smith@", "", "")
+				ca, err := svc.GetClusterAccess("", toolchainv1alpha1.KubesawAuthenticatedUsername, "smith2", "", false)
 
 				// then
-				require.NoError(s.T(), err)
-				require.NotNil(s.T(), ca)
+				require.Error(s.T(), err)
```
Comment (alexeykazakov) [inline pkg/proxy/service/cluster_service_test.go:480]: It's minor but using `require.EqualError()` is usually better IMO. So you can check that you get the exact error you are expecting and not something unrelated to the test case.

**Review by alexeykazakov — APPROVED** [review]
Looks good! Thanks. Just one minor comment.

**Review by mfrancisc — APPROVED** [review]
Great Job 👍 

Thanks for addressing my comments!

**Comment by filariow** [comment]
/retest 

worker setup is failing when downloading Go SDK
```
 go1.20.11.linux-amd64.tar.gz: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match 
```

**Pull request merged**